qcengine/mdi_server.py

```diff
@@ -0,0 +1,408 @@
+    def send_commands(self):
+        """ Send the supported MDI commands through MDI
+
+        :returns: *command_string* String containing the name of each supported command
+        """
+        command_string = "".join([f"{c:{MDI_COMMAND_LENGTH}}" for c in self.commands.keys()])
+
+        # confirm that command_string is the correct length
+        if len(command_string) != len(self.commands) * MDI_COMMAND_LENGTH:
+            raise Exception('Programming error: MDI command_string is incorrect length')
+
+        MDI_Send(command_string, len(command_string), MDI_CHAR, self.comm)
+        return command_string
+
+    # Respond to the <ELEMENTS command
+    def send_elements(self):
+        """ Send the atomic number of each nucleus through MDI
+
+        :returns: *elements* Element of each atom
+        """
+        natom = len(self.molecule.geometry)
+        elements = [qcel.periodictable.to_atomic_number(self.molecule.symbols[iatom]) for iatom in range(natom)]
+        MDI_Send(elements, natom, MDI_INT, self.comm)
+        return elements
+
+    # Respond to the >ELEMENTS command
+    def recv_elements(self, elements=None):
+        """ Receive a set of atomic numbers through MDI and assign them to the atoms in the current molecule
+
+        Arguments:
+            elements: New element numbers. If None, receive through MDI.
+        """
+        natom = len(self.molecule.geometry)
+        if elements is None:
+            elements = MDI_Recv(natom, MDI_INT, self.comm)
+
+        for iatom in range(natom):
+            self.molecule.symbols[iatom] = qcel.periodictable.to_symbol(elements[iatom])
+
+        return elements
+
+    # Respond to the <MASSES command
+    def send_masses(self):
+        """ Send the nuclear masses through MDI
+
+        :returns: *masses* Atomic masses
+        """
+        natom = len(self.molecule.geometry)
+        masses = self.molecule.masses
+        MDI_Send(masses, natom, MDI_DOUBLE_NUMPY, self.comm)
+        return masses
+
+    # Respond to the >MASSES command
+    def recv_masses(self, masses=None):
+        """ Receive a set of nuclear masses through MDI and assign them to the atoms in the current molecule
+
+        Arguments:
+            masses: New nuclear masses. If None, receive through MDI.
+        """
+        natom = len(self.molecule.geometry)
+        if masses is None:
+            masses = MDI_Recv(natom, MDI_DOUBLE, self.comm)
+        self.update_molecule("masses", masses)
+
+    # Respond to the <TOTCHARGE command
+    def send_total_charge(self):
+        """ Send the total system charge through MDI
+
+        :returns: *charge* Total charge of the system
+        """
+        charge = self.molecule.molecular_charge
+        MDI_Send(charge, 1, MDI_DOUBLE, self.comm)
+        return charge
+
+    # Respond to the >TOTCHARGE command
+    def recv_total_charge(self, charge=None):
+        """ Receive the total system charge through MDI
+
+        Arguments:
```

Comment:
Could you use Napoleon style docs here? You can see this style in the associated files.